src/puppetlabs/puppetdb/command.clj

```diff
@@ -86,23 +103,110 @@
 
-;; ## Command parsing
-
-(defn parse-command
-  "Take a wire-format command and parse it into a command object."
-  [{:keys [headers body]}]
+(defn- die-on-header-payload-mismatch
+  [name in-header in-body]
+  (when (and in-header in-body (not= in-header in-body))
+    (throw+
+     (fatality
+      (Exception.
+       (format "%s mismatch between message properties and payload (%s != %s)"
+               name in-header in-body))))))
+
+(def new-message-schema
+  {:headers {:command s/Str
+             :version s/Str
+             :certname s/Str
+             :received s/Str
+             :id s/Str}
+   :body (s/cond-pre s/Str utils/byte-array-class)})
+
+(def queue-message-schema
+  ;; FIXME: do we want to (and can we) be so strict here, i.e. we want
+  ;; to make sure these don't have a command, version, or certname, but
+  ;; this goes further, and isn't appropriate if we'd prefer to
+  ;; pass-through unrelated JMS message headers.
+  ;;
+  ;; Perhaps eventually require :id and :received (cf. PDB-2315).
+  {(s/optional-key :headers) {(s/optional-key :id) s/Str
+                              (s/optional-key :received) s/Str
+                              (s/optional-key :scheduledJobId) s/Str
+                              (s/optional-key :JMSXDeliveryCount) s/Str
+                              s/Any s/Any}
+   :body (s/cond-pre s/Str utils/byte-array-class)})
+
+(def new-command-schema ;; This could probably be stricter
+  {:command s/Str
+   :version s/Int
+   :certname s/Str
+   :payload {s/Any s/Any}
+   :annotations
+   (s/pred
+    (fn [x]
+      (and (map? x)
+           (some-> x :id string?)
+           (some-> x :received string?)
+           (empty? (select-keys x [:command :version :certname])))))})
+
+(def queue-command-schema
+  ;; Created to match existing test behavior, but is that
+  ;; comprehensive? i.e. if we're wrong, and this is too strict, then
+  ;; it could break on outside queue content on release.  Adding a
+  ;; schema here could change our effectively public queue API.
+  (-> new-command-schema
+      ;;(assoc :command (s/cond-pre s/Int s/Str))
+      ;;(assoc :version (s/cond-pre s/Int s/Str))
+      (dissoc :certname)
+      (assoc (s/optional-key :certname) s/Str)
+      (dissoc :payload)
+      (assoc (s/optional-key :payload) (s/cond-pre s/Str {s/Any s/Any}))))
+
+(defn-validated parse-new-command :- new-command-schema
+  "Parses a new-format queue command and returns it in the traditional
+  format (see parse-queue-command).  New-style messages must have all
+  5 headers (command, version, certname, received, and id), and the
+  body must be the bare command content (i.e. the payload).  So for a
+  \"deactivate node\" command, a string like
+  {\"certname\":\"test1\",\"producer_timestamp\":\"2015-01-01\"}."
+  [{:keys [headers body]} :- new-message-schema]
+  (let [{:keys [command version certname received id]} headers
+        version (some-> headers :version Integer/parseInt)]
```

Comment:
this immediately shadows `version` above, and the use of `some->` contradicts the assertion in the docstring that it's required.